cpp/arcticdb/util/enumerate.hpp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+// An internal implementation of folly's enumerate: https://github.com/facebook/folly/blob/main/folly/container/Enumerate.h
+
+#include <iterator>
+#include <arcticdb/util/preprocess.hpp>
+#if defined(_MSC_VER)
+#include <BaseTsd.h>
+typedef SSIZE_T ssize_t;
+#endif
+
+
+namespace arcticdb {
+
+namespace detail {
+
+template <class T>
+struct ConstType {
+    using type = const T;
+};
+
+template <class T>
+struct ConstType<T&> {
+    using type = const T&;
+};
+
+template <class T>
+struct ConstType<T*> {
+    using type = const T*;
+};
+
+template <class Iterator>
+class Enumerator {
+public:
+    constexpr explicit Enumerator(Iterator it) : it_(std::move(it)) {}
+
+    class Proxy {
+    public:
+        using difference_type = ssize_t;
```

Comment:
What is `difference_type` supposed to represent? Is it difference between Enumerator::idx_? If so it seems a bit dangerous as idx_ is of unsigned type.